modules/seqexec/server/src/main/scala/seqexec/server/ObserveActions.scala

```diff
@@ -30,84 +30,86 @@ trait ObserveActions {
+      .ensure(SeqexecFailure.Unexpected("Unable to send DataStart message to ODB."))(identity)
+      .void
+
+  private def sendDataEnd[F[_]: MonadError[?[_], Throwable]](systems: Systems[F], obsId: Observation.Id, imageFileId: ImageFileId, dataId: String): F[Unit] =
+    systems.odb.datasetComplete(obsId, dataId, imageFileId)
+      .ensure(SeqexecFailure.Unexpected("Unable to send DataEnd message to ODB."))(identity)
+      .void
+
+  def observe[F[_]: Sync: Concurrent](
+    systems: Systems[F],
+    config: Config,
+    obsId: Observation.Id,
+    inst: InstrumentSystem[F],
+    otherSys: List[System[F]],
+    headers: Reader[HeaderExtraData, List[Header[F]]])(
+      ctx: HeaderExtraData): Stream[F, Result[F]]
   = {
-    def dataId: SeqActionF[F, String] = SeqActionF.either(
-      config.extractAs[String](OBSERVE_KEY / DATA_LABEL_PROP).leftMap(e =>
-      SeqexecFailure.Unexpected(ConfigUtilOps.explain(e))))
+    def dataId: F[String] =
+      EitherT.fromEither[F](
+        config.extractAs[String](OBSERVE_KEY / DATA_LABEL_PROP)
+          .leftMap(e => SeqexecFailure.Unexpected(ConfigUtilOps.explain(e)))
+      ).widenRethrowT
 
-    def notifyObserveStart: SeqActionF[F, Unit] =
+    def notifyObserveStart: F[Unit] =
       otherSys.map(_.notifyObserveStart).sequence.void
 
     // endObserve must be sent to the instrument too.
-    def notifyObserveEnd: SeqActionF[F, Unit] =
+    def notifyObserveEnd: F[Unit] =
       (inst +: otherSys).map(_.notifyObserveEnd).sequence.void
 
-    def closeImage(id: ImageFileId): SeqActionF[F, Unit] =
-      SeqActionF.embedF(inst.keywordsClient.closeImage(id))
+    def closeImage(id: ImageFileId): F[Unit] =
+      inst.keywordsClient.closeImage(id)
 
-    def doObserve(fileId: ImageFileId): SeqActionF[F, Result[F]] =
+    def doObserve(fileId: ImageFileId): F[Result[F]] =
       for {
         d   <- dataId
         _   <- sendDataStart(systems, obsId, fileId, d)
         _   <- notifyObserveStart
-        _   <- SeqActionF.embedF(headers(ctx).map(_.sendBefore(obsId, fileId)).sequence)
+        _   <- headers(ctx).map(_.sendBefore(obsId, fileId)).sequence
         _   <- info(s"Start ${inst.resource.show} observation ${obsId.format} with label $fileId")
         r   <- inst.observe(config)(fileId)
         _   <- info(s"Completed ${inst.resource.show} observation ${obsId.format} with label $fileId")
         ret <- observeTail(fileId, d)(r)
       } yield ret
 
-    def observeTail(id: ImageFileId, dataId: String)(r: ObserveCommandResult): SeqActionF[F, Result[F]] = {
-      def okTail(stopped: Boolean): SeqActionF[F, Result[F]] = for {
+    def observeTail(id: ImageFileId, dataId: String)(r: ObserveCommandResult): F[Result[F]] = {
+      def okTail(stopped: Boolean): F[Result[F]] = for {
         _ <- notifyObserveEnd
-        _ <- SeqActionF.embedF(headers(ctx).reverseMap(_.sendAfter(id)).sequence.void)
+        _ <- headers(ctx).reverseMap(_.sendAfter(id)).sequence.void
```

Comment:
What is `reverseMap`?

Reply:
It is part of the std library. Like `map` but iterates in reverse